src/utils/async.ts

```diff
@@ -48,3 +48,109 @@ export async function asyncIterableToArray<T>(
   const iter = source[Symbol.asyncIterator]();
   return asyncIteratorToArray(iter);
 }
+
+/**
+ * Create a new AsyncIterable providing only the values that satisfy the predicate function.
+ * The predicate function is also an optional Type Guard function if types T and K are different.
+ *
+ * Example:
+ * ```
+ * const i: AsyncIterable<Either<E, A>> = {} as AsyncIterable<Either<E, A>>;
+ * const f: AsyncIterable<Right<E, A>> = filterAsyncIterable<Either<E, A>, Right<E, A>>(i, isRight);
+ * ```
+ * @param iterable Original AsyncIterable
+ * @param predicate Predicate function
+ */
+export const filterAsyncIterable = <T, K = T>(
+  iterable: AsyncIterable<T | K>,
+  predicate: (value: T | K) => value is K
```

Comment:
How does this work when value is not a union of more than one type? Say we have an AsyncIterable<string> and the predicate returns true for non empty strings, what are T and K?

Reply:
It works with any input type, even if not an union. For example, this code transcompile successfully
```typescript
const i: AsyncIterator<unknown> = {} as AsyncIterator<unknown>;
  const newI: AsyncIterator<NonEmptyString> = filterAsyncIterator<unknown, NonEmptyString>(i, (_): _ is NonEmptyString => {
    if (isRight(NonEmptyString.decode(_))) {
      return true;
    }
    return false;
});
```

Reply:
glass half-full: this is unpractical, but it would force a more fine-grained modeling 
```ts
const RetrievedPendingMessage = t.intersection([
  RetrievedMessage,
  t.interface({ isPending: t.literal(true) })
]);

messageModel
    .findMessages(fiscalCode)
    .map(flattenAsyncIterator)
    .map(_ => filterAsyncIterator(_, RetrievedPendingMessage.is))
```

Reply:
I realised the above code to be wrong. In 
```ts
.map(_ => filterAsyncIterator(_, RetrievedPendingMessage.is))
```
`_` is of type `AsyncIterator<Either<t.Errors, RetrievedMessage>>`, thus the predicate will take an Either.

Should `filterAsyncIterator` care about the value to be `left` or `right`?

Reply:
@balanza yes, you have an Either in input. the right code will be:
```typescript
messageModel
  .findMessages(fiscalCode)
  .map(flattenAsyncIterator)
  .map(_ =>
    filterAsyncIterator<
      Either<Errors, RetrievedMessage>,
      Right<Errors, RetrievedPendingMessage>
    >(
      _,
      (val): val is Right<Errors, RetrievedPendingMessage> =>
        isRight(val) && RetrievedPendingMessage.is(val)
    )
  );
```

Reply:
Maybe this signature is more compact and clear:
```typescript
export function filterAsyncIterator<T>(
  iter: AsyncIterator<unknown>,
  predicate: (value: unknown) => value is T
): AsyncIterator<T> {
```

Reply:
we probably need two signatures to support both non type-guards and type guards predicates, as in the standard library: https://github.com/microsoft/TypeScript/blob/master/lib/lib.es5.d.ts#L1176